source/Ship.cpp

```diff
@@ -1721,6 +1721,9 @@ void Ship::Move(vector<Visual> &visuals, list<shared_ptr<Flotsam>> &flotsam)
 			double cost = attributes.Get("turning energy");
 			if(energy < cost * fabs(commands.Turn()))
 				commands.SetTurn(commands.Turn() * energy / (cost * fabs(commands.Turn())));
+			cost = attributes.Get("turning fuel");
+			if(fuel < cost * fabs(commands.Turn()))
+				commands.SetTurn(commands.Turn() * fuel / (cost * fabs(commands.Turn())));
```

Comment:
I don't think this will work out.

What this is currently doing is limiting the turn rate by the energy cost, *then* by the fuel cost, instead of limiting by both at once. So say you require 10 energy and 10 fuel but you only have 1 energy and 5 fuel available. What *should* happen is that you use 1 energy and 1 fuel for 10% turn power as that's what you have available. I believe that what this is doing is seeing that you only have enough energy for 10% turn power and only enough fuel for 50% turn power, so it'd stack those two limits and give you 5% power because of you're calling SetTurn twice and the second call is getting the turn rate from the previous call because of the Turn call.

What you should be doing is checking your available energy and fuel, then taking the minimum of the two to set your turn. And while we're here, I think it would also make sense to allow heat to be used as a resource (i.e. check if heat + turn heat > 0 and limit the turn if it isn't), as we've done that with weapons and repairs.

This same issue appears to be the case for thrusting.

Reply:
That's a good point, heat-wise, though I believe because we're decreasing "cost" by our available energy, it'll only check if the amount of thrust/turn we can produce energy-wise will violate our fuel usage too, e.g.

If we use 10 energy and 10 fuel, and we have 1 energy and 5 fuel, we check to see if 10>1, and set our usage to 10%. Then, we check if 10*0.1>5, and see that since we're consuming only 1 fuel out of a total of 5, things check out. The same sort of deal going on in DoRepair(), but it's been heavily streamlined and generalized. The afterburner situation is different, and sort of assumes that one frame is the indivisible unit of afterburner production, require enough fuel ans energy for a full frame of full-blast afterburner.

Tomorrow I'll check for negative thrusting/turning heat and only allow movement if your heat will remain >= 0